File: Type.hs
-- Haskell Example 2
-- Type inference

-- the type of foo is inferred to be Integer -> Integer -> Integer  
-- since performing a '+' operation on a number literal must mean that x is an integer
-- Type inference is done at compile time in Haskell
foo x = x + 1

main = do
    putStrLn "5 + 1: "
    print(foo 5)

File: Type.java
// Java Example 2
// Type inference

public class Type {
    // a return type of 'int' is required in the function declaration 
    // it cannot be inferred by the compiler in Java
    public static int foo(int x) {
        return x + 1;
    }

    public static void main(String[] args) {
        System.out.println(foo(5));
    }
}
